Allow network-ntt implementation backend in exe-commons

qnikst opened the pull request.

mzabaluev commented on exe-common/src/config.rs:
```diff
@@ -49,13 +49,16 @@ pub mod net {
         pub fn new(addr: String) -> Self {
             if addr.starts_with(r"http://") || addr.starts_with(r"https://") {
                 Peer::http(addr)
+            } else if addr.starts_with(r"ntt://") {
```
Eek, vendors should not invent URI schemes just like that. Use `x-cardano-ntt` perhaps.

NicolasDP replied:
that's a good point. Actually what about using multiaddr?

mzabaluev commented on exe-common/src/lib.rs:
```diff
@@ -1,3 +1,4 @@
+#![type_length_limit = "2097152"]
```
I wonder what monster type causes this.
What's the error if this directive is removed?

NicolasDP replied:
holy 🐮 ! we have a type that needs 2 MB ?

qnikst replied:
Hurray after a number of rewrites of the ntt-client, this requirement has gone.

mzabaluev commented on exe-common/src/network/ntt.rs:
```diff
@@ -0,0 +1,73 @@
+    tx::{TxAux, TxId},
+};
+
+pub struct NetworkCore {
```
Why is this named `NetworkCore`?

qnikst replied:
I was planning to have a single structure that will keep runtime for all the threads, but that didn't work out. Any better name suggestion will be appreciated.

mzabaluev commented on exe-common/src/network/ntt.rs:
```diff
@@ -0,0 +1,73 @@
+    }
+}
+
+impl Api for NetworkCore {
```
Ugh, why do we have this trait named `Api` to begin with. Not a problem of this PR, of course.

NicolasDP replied:
old exe-common things...

mzabaluev commented on exe-common/src/network/ntt.rs:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    fn wait_for_new_tip(&mut self, _prev_tip: &HeaderHash) -> Result<BlockHeader> {
+        panic!("not yet ready")
```
Could be just `unimplemented!()`

mzabaluev commented on exe-common/src/network/peer.rs:
```diff
@@ -29,6 +31,16 @@ impl Peer {
+                match addrs_iter.next() {
+                    Some(addr) => ntt::NetworkCore::new(addr).map(Peer::Ntt),
+                    None => Err(Error::from(std::io::Error::new(
+                        std::io::ErrorKind::Other,
```
I would say `ErrorKind::InvalidInput` fits better here. But even better is to have a case for this in `network::Error` rather than go through `std::io::Error`.

qnikst replied:
I'm not sure I want to propagate all the errors in the rust-cardano, I'm not sure I'm ready for that endeavour.

mzabaluev replied:
In fact, what's needed here is handling the error return of `to_socket_addrs()` above instead of dying on an invalid address.

Consider adding a variant `InvalidPeerAddress(String)` to `network::Error`; the address string that failed parsing goes inside the value.

qnikst replied:
Fixed

mzabaluev commented on network-grpc/src/client.rs:
```diff
@@ -1,6 +1,6 @@
 use crate::gen::{self, node::client as gen_client};
 
-use chain_core::property::{self, Deserialize, FromStr, HasHeader, Serialize};
+use chain_core::property::{self, Block, Deserialize, FromStr, HasHeader, Serialize};
```
Rebase this as well to eliminate cross-pollination from other branches. I'll continue reviewing from the clean branch.

qnikst commented:
Please take another look, I've addressed all the comments except for naming and possibly multiaddr.
Currently the only problem I see is that I'm creating a runtime for each connection, ideally, I should create only one runtime for all the program, but I'm not sure how to fit it here.

mzabaluev commented on protocol-tokio/src/protocol/codec/handshake.rs:
```diff
@@ -43,6 +43,12 @@ impl de::Deserialize for ProtocolMagic {
     }
 }
 
+impl From<&u32> for ProtocolMagic {
```
Why does it need to be a reference? `u32` is copyable.

qnikst replied:
Fixed

mzabaluev commented on exe-common/src/config.rs:
```diff
@@ -49,13 +49,16 @@ pub mod net {
             if addr.starts_with(r"http://") || addr.starts_with(r"https://") {
                 Peer::http(addr)
+            } else if addr.starts_with(r"x-cardano-ntt://") {
+                Peer::ntt(addr[17..].to_string())
```
Could use .len() instead of counting bytes.

qnikst replied:
fixed

mzabaluev commented on exe-common/src/network/peer.rs:
```diff
@@ -29,6 +31,13 @@ impl Peer {
                 Ok(Peer::Http(hermes::HermesEndPoint::new(addr, network)))
             }
+            config::net::Peer::Ntt(addr) => {
+                let mut addrs_iter = addr.to_socket_addrs().unwrap();
```
The `unwrap` is still there, meaning that jormungandr will panic on any invalid address?

qnikst replied:
I hope this code is never used in jormundandr.. but you are right that cardano-cli will fail

mzabaluev replied:
Sorry, I meant cardano-cli. It should be easy to check for the error here and convert it to the same `InvalidPeerAddress` error variant.

qnikst replied:
Fixed

mzabaluev replied:
Hmm, the `unwrap` is still there?

mzabaluev commented on exe-common/src/network/error.rs:
```diff
@@ -66,6 +67,7 @@ impl fmt::Display for Error {
             Error::NoSuchBlock(hash) => write!(f, "Requested block {} does not exist", hash),
             Error::StorageError(_) => write!(f, "Storage error"),
             Error::BlockError(_) => write!(f, "Block error"),
+            Error::InvalidPeerAddress(addr) => write!(f, "Block error {}", addr),
```
Something like `"Invalid peer address '{}'", addr`, perhaps?

qnikst replied:
Oh.. my mistake, thanks, fixed.

mzabaluev commented on exe-common/src/config.rs:
```diff
@@ -49,13 +49,16 @@ pub mod net {
             if addr.starts_with(r"http://") || addr.starts_with(r"https://") {
                 Peer::http(addr)
+            } else if addr.starts_with(r"x-cardano-ntt://") {
+                Peer::ntt(addr["x-cardano-ntt://".len()..].to_string())
```
OK, though using a string constant would be better.

mzabaluev replied:
It's also not an `r`-literal, though I'm not sure there is a practical difference. Small things...

mzabaluev commented:
This is good to go after a touch of rustfmt.
